custom_types.md

```diff
@@ -0,0 +1,54 @@
+# Custom types
+
+Custom types is a gogo protobuf extensions that allows for using a custom
+struct type to decorate the underlying structure of the protocol message.
+
+# How to use
+
+## Defining the protobuf message
+
+``` protocol-buffer
+message CustomType {
+  optional ProtoType Field = 1 [(gogoproto.customtype) = "GoType"];
+}
+
+message ProtoType {
+  optional string Field = 1;
+}
+```
+
+or alternatively you can declare the field type in the protocol message to be
+`bytes`:
+
+``` protocol-buffer
+message BytesCustomType {
+  optional bytes Field = 1 [(gogoproto.customtype) = "GoType"];
+}
+
+message ProtoType {
+  optional string Field = 1;
+}
+
+```
+
+The downside of using `bytes` is that it makes it harder to generate protobuf
+code in other languages. In either case, it is the user responsibility to
+ensure that the custom type marshals and unmarshals to the expected wire
+format. That is, in the first example, gogo protobuf will not attempt to ensure
+that the wire format of `ProtoType` and `GoType` are wire compatible.
+
+## Custom type method signatures
+
+The custom type must define the following methods with the given
+signatures. Assuming the custom type is called `GoType`:
+
+``` go
+func (r GoType) Marshal() ([]byte, error)
+func (r *GoType) MarshalTo(data []byte) (n int, err error)
+func (r *GoType) Unmarshal(data []byte) error
+
+func (gt GoType) MarshalJSON() ([]byte, error)
+func (gt *GoType) UnmarshalJSON(data []byte) error
+
+func (gt GoType) Compare(other GoType) int // only required if the comparable option is set
+```
```

Comment:
What about Equal and NewPopulated, etc.
I think the comment about a method being optional can be added above the method.
You can also get the highlighting to keep on working by simply adding curly braces to the function definitions.
```go
func (r GoType) Marshal() ([]byte, error) {}
```